README.md
# Web server for HDL Buspro
**This is a web server for setting up and communicating with the smart home system *HDL Buspro*. It was mainly built for
the purpose of integrating with *Google Assistant* (Google Home and Google Pixel).**

## Table of Contents
- [Video demo](#video-demo)
- [How it works](#how-it-works)
- [Example commands](#example-commands)
- [Running the server - quick and dirty](#running-the-server---quick-and-dirty)
- [Running the server - complete setup on a Raspberry Pi](#running-the-server---complete-setup-on-a-raspberry-pi-running-raspbian-gnulinux)
- [Security](#security)
- [Links](#links)

## Video demo
[![YouTube proof of concept video](http://img.youtube.com/vi/P6BgsM8IiW8/0.jpg)](https://youtu.be/P6BgsM8IiW8)

## How it works
Google have added some [triggers at IFTTT](https://ifttt.com/google_assistant) for Google Assistant (i.e. Google Home
and Google Pixel). Combining them with the [Maker actions](https://ifttt.com/maker), a web request can be sent to the
HDL Buspro server in this project. The server interprets the requests by using the actions you have set up and sends the
belonging HDL commands to the IP gateway.

## Example commands
 - *Dim the lights to [X]% in [area/room]*
 - *Set the lights to [X]% in [area/room]*
 - *Turn [on/off] the lights in [area/room]*
 - *Turn [on/off] the TV*
 - *[Open/close] the garage*
 - *Bye bye*
 - *Turn [on/off] the alarm*
 - *Send for help*

## Running the server - quick and dirty
Prerequisites:
 - [Java 8](https://java.com/en/download/)
 - [The server JAR](https://github.com/roys/java-hdl-webserver/raw/master/releases/hdlserver-LATEST.jar)

### Un*x
To start the server on Linux, Mac and other Unix-like systems:  
`nohup java -cp hdlserver-LATEST.jar com.roysolberg.java.hdlserver.Application &`

The server will start logging to the `logs` directory. See that everything is running fine:
`tail -f logs/hdl-server.log`

### Windows
To start the server on Windows:  
`javaw -cp hdlserver-LATEST.jar com.roysolberg.java.hdlserver.Application`  
Alternatively:  
`start java -cp hdlserver-LATEST.jar com.roysolberg.java.hdlserver.Application`

See that everything is fine by looking in the file `logs\hdl-server.log`.

### Un*x + Windows
You should now be able to go to the server in a browser from a computer on the local network. The server is running on port 4567, so go to `http://<server ip address>:4567`.

Next step is to [create some actions](#part-2-adding-commands-actions).

## Running the server - complete setup on a Raspberry Pi running Raspbian GNU/Linux

### Part 1: Download and set up the server

#### 1. Java setup

First SSH into your Raspberry Pi.

Type `java -version` to see which Java version you have. If it says something with `1.8` or later you are good to go. Just continue to the next step.

If you don't have Java installed or a Java version before version 1.8, run the following commands and follow the setup:  
`sudo apt-get update`  
`sudo apt-get upgrade`  
`sudo apt install default-jdk` (old command: ~~`sudo apt-get install oracle-java8-jdk`~~)  

#### 2. Install Git

If you don't have the version-control system Git installed you can install it like this:  
`sudo apt-get install git`  

#### 3. Directory setup and getting the server

First check out the code:  
`cd ~`  
`git clone https://github.com/roys/java-hdl-webserver.git`  

It can be smart to check that the server is working and getting data:  
`cd java-hdl-webserver/releases`  
`java -cp hdlserver-LATEST.jar com.roysolberg.java.hdlserver.Application`

You should see some output confirming that the server is starting up ok. Please note that it could take a few seconds to start up on hardware like one of the Raspberry Pi models.

Use `ctrl + c` to exit the application.

#### 4a. Running the server in the background

If you want to make the server run on startup then skip to **4b**. If you want to just run the server in the background (even after you log out) you can enter the following
command:  
`nohup java -cp hdlserver-LATEST.jar com.roysolberg.java.hdlserver.Application &`  

The number that is printed out is the process number. You can use this number to stop the script.

If you don't have the process number you can find it using `ps aux | grep hdlserver`.

To kill/stop the script just enter the following command:
`kill [process number]`  

#### 4b. Running the server on startup

The server can be run as a cron job that starts on reboot of the Raspberry Pi.

To edit the cron jobs:  
`crontab -e`

Add the following line at the end:  
`@reboot /bin/sh ~/java-hdl-webserver/launcher.sh`

Use `ctrl + x` to exit, press `Y` to save the cron setup and `enter` to confirm the filename.

Restart the device using `sudo shutdown -r now` and check out if there's stuff going on in `releases/logs/hdl-bus.log`. `releases/logs/hdl-server.log` and `releases/logs/nohup.out` will also contain output and any error messages running the server. Please note that there is a delay of 30 seconds when running the server from the `launcher.sh` script (to ensure that network interfaces are up and IP address is assigned, etc).

#### 5. Reading the logs

You can easily tap in to what's going by running `tail`:  
`tail -f hdl-bus.log`  
`tail -f hdl-server.log`  

This will let you follow the file and the contents that's written to it. To end the session just press `ctrl + c`.

#### 6. Remote access

To be able to have IFTTT access (described later) the server and run the commands you set up you have to ensure that the server is accessible from the Internet. How to do this depends on your Internet router, but in general it will be like this:
 - Give the computer running the server a static IP address in your router
 - Forward a port (please use some other high number port than 4567) from Internet to port 4567 to the computer running the server
 - Set up some kind of dynamic DNS (can often be done on the router itself, or you can use the computer running the server) to get a static hostname for IFTTT
 - Use the hostname and external port number when setting up IFTTT (described later)

#### 7. Up and running
You should now be able to go to the server in a browser from a computer on the local network. The server is running on port 4567, so go to `http://<server ip address>:4567`.

Next step is to [create some actions](#part-2-adding-commands-actions).

#### Updating the server

If you chose to do a `git clone` in the first step you can just run `git pull` inside the directory at `~/java-hdl-webserver`. If you run the script at boot
time it is easiest to just reboot your Raspberry Pi using `sudo shutdown -r now` to ensure the latest version of the script is running.

### Part 2: Adding commands (actions)

Prerequisites: An account at https://ifttt.com and a Google account.

#### Create dimmer actions

Dimmer actions let you say commands like *Dim the lights to 20% in the kitchen* and are meant for HDL dimmers.

<img align="right" src="https://github.com/roys/java-hdl-webserver/raw/master/misc/01-create_dimmer_action.png" alt="Example of creating dimmer action for HDL Buspro server" title="Example of creating dimmer action for HDL Buspro server" width="500" />

1. Go to to the **Actions** page.
2. Below **Create dimmer action**, type in the name of the area. This will be the area you asked to be dimmed, e.g. *The kitchen*.  
You can also add zero to many aliases for the area if you use different names for it. If using *"the"* in an area name you might want to repeat the name of the area without *"the"*. Then everything will work if you say *Dim the lights to 20% in kitchen*.
3. Select the dimmer channel representing the area.  
If you want the same area to control several areas or you have several channels for the same area you can add those as well.
4. Select *Create action* and notice the info about the action at the bottom of the page.
5. Go to https://ifttt.com/create to create a new "applet" (preferably in a new browser window/tab).
6. Select the service *Google Assistant*.
7. Select *Say a phrase with both a number and a text ingredient*.
8. Choose up to three commands you want to say to trigger the action, e.g.:
 - *Dim the lights to # % in $*
 - *Turn up the lights to # % in $*
 - *Set the lights to # % in $*
9. Choose a response from the Assistant. It can be nice to include the input from your command, e.g. *OK, # % in $*.
10. Create the trigger.
11. Select the action service *Maker*.
12. Select *Make a web request*.
13. Fill in the URL, method (POST), content type (application/json) and body as described in the **Actions** page in step 4.
14. Create the action.
15. Don't be to impatient, but you should now be able to test out the command you just created, e.g. *OK Google, dim the lights to 20% in the kitchen*.

#### Create custom actions

Custom actions let you give custom commands like *Open the garage* which can trigger some panel button or set a universal switch.

<img align="right" src="https://github.com/roys/java-hdl-webserver/raw/master/misc/02-create_custom_action.png" alt="Example of creating custom action for HDL Buspro server" title="Example of creating custom action for HDL Buspro server" width="500" />

1. Go to to the **Actions** page.
2. Below **Create custom action**, type in the name/description of the command.  
This command is just a description and doesn't have to be the same as the voice command.
3. Select the component, operation, function and parameters.  
**Panel control** in combination with **Control button** lets you specify the button number as parameters 2 and 1 for on and 0 for off.  
If you want the same command to do more commands you can add them.
4. Select *Create action* and notice the info about the action at the bottom of the page.  
5. Go to https://ifttt.com/create to create a new "applet" (preferably in a new browser window/tab).
6. Select the service *Google Assistant*.
7. Select *Say a simple phrase*.
8. Choose up to three commands you want to say to trigger the action, e.g. :
 - *Open the garage*
 - *Close the garage*
 - *Open the garage, please*
9. Choose a response from the Assistant.
10. Create the trigger.
11. Select the action service *Maker*.
12. Select *Make a web request*.
13. Fill in the URL, method (POST), content type (application/json) and body as described in the **Actions** page in step 4.
14. Create the action.
15. Don't be to impatient, but you should now be able to test out the command you just created, e.g. *OK Google, open the garage*.

**Unlike the dimmer actions you need to add both a new action and a new "applet" if you want to create another one.**

## Security
I strongly recommend using HTTPS for the server. Setting that up is a bit out of scope for this document, but you should
really look into it. If not using HTTPS someone on the Internet might pick up the commands you are sending. The commands
can replayed and someone might guess how to use other commands.

The reason there is no login for the frontend of the web server is simple: The moment someone has access to the network
where there are HDL Buspro traffic they don't need this server. All communication is done in plain text with no
security. Therefore securing the frontend with a login would add a false sense of security. **The web server front end
(not the API for the commands) requires the user to be in the same local network.**

The security for the API and the commands lies in an automatically created random key when starting the server the first
time. If the requests for the commands are missing the correct key they will not be executed.

Keep in mind that there could potentially be security holes or weaknesses in the code of this project, the web framework this is built
on, the built-in application server this runs on, the server you are running, your router, other network components,
IFTTT and Google Assistant. You have no guarantees. If you want to stay safe, you should stay offline.

If you want to learn a little more about security and HDL Buspro you read my Google+ post at
https://blog.roysolberg.com/2017/08/my-dumb-smart-home .

## Links
 - This repo: https://github.com/roys/java-hdl-webserver
 - My Android app for controlling HDL Buspro: https://play.google.com/store/apps/details?id=com.roysolberg.android.smarthome
 - YouTube demo: https://youtu.be/P6BgsM8IiW8
 - HDL Buspro: http://www.hdlautomation.com
 - Google Assistant: https://assistant.google.com
 - Google Home: https://madeby.google.com/home/
 - Google Assistant at IFTTT: https://ifttt.com/google_assistant
 - Maker at IFTTT: https://ifttt.com/maker


releases/CHANGELOG.txt
Changelog for HDL Buspro server

List of changes, newest changes first.
('+' = new feature, '-' = removed feature, '*' = changed feature)

Version 1.0.3
 * Upgraded from using Spark version 2.5.3 to version 2.8.0. See https://github.com/advisories/GHSA-76qr-mmh8-cp8f .
 * Fixed now broken link.
Version 1.0.2
 + Added support for curtain switch control (0xE3E0) in custom action.
Version 1.0.1
 + Added support for single channel control (0x0031) in custom action.
 * Bugfix: Deletion of actions did not work properly.
 * Authentication key only has to be 25 characters long.
Version 1.0.0
 + Initial release.


src/main/java/com/roysolberg/java/hdlserver/Application.java
package com.roysolberg.java.hdlserver;

import com.roysolberg.java.hdlserver.hdl.Action;
import com.roysolberg.java.hdlserver.hdl.component.HdlComponent;
import com.roysolberg.java.hdlserver.service.HdlService;
import com.roysolberg.java.hdlserver.util.HdlUtil;
import org.apache.commons.lang.StringEscapeUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.mapdb.DB;
import org.mapdb.DBMaker;
import spark.ModelAndView;
import spark.Request;
import spark.template.velocity.VelocityTemplateEngine;

import java.io.UnsupportedEncodingException;
import java.math.BigInteger;
import java.net.InetAddress;
import java.net.URLEncoder;
import java.net.UnknownHostException;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentMap;

import static spark.Spark.*;

public class Application {

    private static Logger logger = LogManager.getLogger(Application.class);

    protected final static boolean DEVEL_MODE = false;

    protected HdlService hdlService;
    protected DB database;
    protected ConcurrentMap configConcurrentMap;
    protected ConcurrentMap componentsConcurrentMap;
    protected ConcurrentMap actionsConcurrentMap;
    protected ConcurrentMap actionAliasesConcurrentMap;
    protected List<HdlComponent> hdlComponents;

    public Application() {
        setUpDatabase();
        setUpConfig();
        setUpHdlService();
        setUpServer();
        setUpPaths();
    }

    public static void main(String[] args) {
        new Application();
    }

    protected void setUpDatabase() {
        database = DBMaker.fileDB("hdlserver.db").fileMmapEnable().checksumHeaderBypass().make();
    }

    protected void setUpConfig() {
        configConcurrentMap = database.hashMap("config").createOrOpen();
        if (!configConcurrentMap.containsKey("authToken")) {
            configConcurrentMap.put("authToken", generateAuthToken());
            database.commit();
        }
        componentsConcurrentMap = database.hashMap("components").createOrOpen();
        actionsConcurrentMap = database.hashMap("actions").createOrOpen();
        actionAliasesConcurrentMap = database.hashMap("actionAliases").createOrOpen();
    }

    protected String generateAuthToken() {
        return new BigInteger(130, new SecureRandom()).toString(32).toUpperCase();
    }

    protected void setUpHdlService() {
        hdlService = new HdlService(database);
    }

    protected void setUpServer() {
        if (DEVEL_MODE) {
            logger.info("***");
            logger.info("*** RUNNING IN DEVELOPMENT MODE");
            logger.info("***");
            staticFiles.externalLocation(System.getProperty("user.dir") + "/src/main/resources/public");
            System.out.println("System.getProperty(\"user.dir\"):" + System.getProperty("user.dir"));
        } else {
            staticFiles.location("/public");
            staticFiles.expireTime(600);
        }
    }

    // TODO: Add status page for HDL service, and settings page for changing auth token and somewhere to see the performed commands and other interesting events
    protected void setUpPaths() {
        before((request, response) -> {
            logger.info("Request: {} - \"{} {} {}\" - \"{}\"", request.ip(), request.requestMethod(), request.pathInfo(), request.protocol(), request.userAgent());
        });
        get("/", (req, res) -> {
            requireSiteLocalAddress(req);
            return getModelAndView(req, "index");
        }, new VelocityTemplateEngine());
        get("/actions", (req, res) -> {
            requireSiteLocalAddress(req);
            return getModelAndView(req, "actions");
        }, new VelocityTemplateEngine());
        post("/create_custom_action", (request, response) -> {
            requireSiteLocalAddress(request);
            // TODO: Error handling and success + error feedback
            createCustomAction(request.queryParams("description"), request.queryParamsValues("components[]"), request.queryParamsValues("operations[]"), request.queryParamsValues("parameters1[]"), request.queryParamsValues("parameters2[]"), request.queryParamsValues("parameters3[]"));
            response.redirect("/actions");
            return null;
        });
        post("/create_dimmer_action", (request, response) -> {
            requireSiteLocalAddress(request);
            // TODO: Error handling and success + error feedback
            createDimmerAction(request.queryParams("description"), request.queryParamsValues("aliases[]"), request.queryParamsValues("channels[]"));
            response.redirect("/actions");
            return null;
        });
        get("/security", (req, res) -> {
            requireSiteLocalAddress(req);
            return getModelAndView(req, "security");
        }, new VelocityTemplateEngine());
        get("/demo", (req, res) -> {
            requireSiteLocalAddress(req);
            return getModelAndView(req, "demo");
        }, new VelocityTemplateEngine());
        get("/settings", (req, res) -> {
            requireSiteLocalAddress(req);
            return getModelAndView(req, "settings");
        }, new VelocityTemplateEngine());
        post("/api/actions/:actionId", "application/json", (request, response) -> {
            JSONObject jsonObject = (JSONObject) new JSONParser().parse(request.body());
            requireValidAuthToken((String) jsonObject.get("auth"));

            String actionId = request.params(":actionId");

            Action action = null;
            printRequestDebugInfo(request);

            try {
                action = (Action) actionsConcurrentMap.get(Integer.parseInt(actionId));
                if (action.getActionType() != Action.ActionType.Custom) {
                    action = null;
                }
            } catch (NumberFormatException e) {
                logger.error("Got NumberFormatException while trying to parse action id [" + actionId + "]. Sending back HTTP 404.", e);
            }

            if (action == null) {
                halt(404, "Action not found.");
            }

            // TODO: Log all performed actions
            hdlService.performDimmerAction(action);

            response.status(202);
            response.type("application/json");
            return "{\"status\":\"ok\"}";
        });
        post("/api/actions/dimmer/:alias", "application/json", (request, response) -> {
            response.type("application/json");
            Action action = null;
            int valueInPercent = 0;
            try {
                String rawJson = request.body();
                logger.info("Actions:" + actionsConcurrentMap.keySet());
                logger.info("Aliases:" + actionAliasesConcurrentMap.keySet());
                logger.info("Raw json: [" + rawJson + "]");
                JSONObject jsonObject = (JSONObject) new JSONParser().parse(rawJson);
                requireValidAuthToken((String) jsonObject.get("auth"));

                String areaAlias = request.params(":alias");
                logger.info("Area1: [" + areaAlias + "]");
                areaAlias = areaAlias.trim().toLowerCase();
                logger.info("Area2: [" + areaAlias + "]");

                valueInPercent = ((Long) jsonObject.get("value")).intValue();

                printRequestDebugInfo(request);

                Object actionId = actionAliasesConcurrentMap.get(areaAlias);
                if (actionId != null) {
                    action = (Action) actionsConcurrentMap.get(actionId);
                    logger.info("Found action [" + action + "].");
                }
            } catch (Exception e) {
                e.printStackTrace();
                logger.error("Got exception while handling request.", e);
                halt(500, "{\"status\":\"error\"}");
            }
            if (action == null) {
                halt(404, "Action not found.");
            }
            // TODO: Log all performed actions
            hdlService.performDimmerAction(action, valueInPercent);

            response.status(202);
            return "{\"status\":\"ok\"}";
        });
        post("/delete_action", (request, response) -> {
            String successMessage = null;
            String errorMessage = null;
            String actionId = request.queryParams("actionId");
            logger.info("Actions:" + actionsConcurrentMap.keySet());
            logger.info("actionId:" + actionId);
            try {
                Action deletedAction = (Action) actionsConcurrentMap.remove(Integer.parseInt(actionId));
                if (deletedAction != null) {
                    if (deletedAction.getAliases() != null) {
                        actionAliasesConcurrentMap.remove(deletedAction.getDescription());
                        for (String alias : deletedAction.getAliases()) {
                            actionAliasesConcurrentMap.remove(alias);
                        }
                    }
                    successMessage = "Action was successfully deleted.";
                } else {
                    errorMessage = "Unable to find action.";
                }
            } catch (NumberFormatException e) {
                logger.error("Got NumberFormatException while trying to parse action id [" + actionId + "]. Sending back HTTP 404.", e);
            }
            response.redirect("/actions?" + getMessagesQueryParams(successMessage, errorMessage));
            return null;
        });
        post("/search_for_components", (request, response) -> {
            hdlService.doOneTimeQuickSearchForComponents();
            String successMessage = "Performing search. Please refresh page to see any new components found.";
            String errorMessage = null;
            response.redirect("/?" + getMessagesQueryParams(successMessage, errorMessage));
            return null;
        });
        post("/update_settings", (request, response) -> {
            String successMessage = null;
            String errorMessage = null;
            String authKey = request.queryParams("authKey");
            if (authKey != null) {
                authKey.trim();
                if (authKey.length() >= 25) {
                    configConcurrentMap.put("authToken", authKey);
                    database.commit();
                    successMessage = "Settings successfully updated";
                } else {
                    errorMessage = "Authorization key must be at least 26 characters.";
                }
            } else {
                errorMessage = "Authorization key must be at least 26 characters.";
            }

            response.redirect("/settings?" + getMessagesQueryParams(successMessage, errorMessage));
            return null;
        });

    }

    protected String getMessagesQueryParams(String successMessage, String errorMessage) throws UnsupportedEncodingException {
        String query = "";
        if (successMessage != null) {
            query += "&success=" + URLEncoder.encode(successMessage, "UTF-8");
        }
        if (errorMessage != null) {
            query += "&error=" + URLEncoder.encode(errorMessage, "UTF-8");
        }
        return query;
    }

    protected void printRequestDebugInfo(Request request) {
        //if (DEVEL_MODE) {
        System.out.println("Host: " + request.host());
        logger.info("Host: " + request.host());
        System.out.println("IP: " + request.ip());
        logger.info("IP: " + request.ip());
        for (String headerKey : request.headers()) {
            String log = headerKey + "=" + request.headers(headerKey);
            System.out.println(log);
            logger.info(log);
        }
        //}
    }

    protected boolean createCustomAction(String description, String[] componentIds, String[] operations, String[] parameters1, String[] parameters2, String[] parameters3) {
        // TODO: Improve error handling, logging and user feedback
        if (componentIds != null && componentIds.length > 0 && operations != null && operations.length == componentIds.length) { // At least one component + operation
            // TODO: Improve action ID generation:
            int actionId = actionsConcurrentMap.size() + 1;
            while (actionsConcurrentMap.containsKey(actionId)) {
                actionId++;
            }
            Action action = new Action(Action.ActionType.Custom, actionId, description);
            for (int i = 0; i < componentIds.length; i++) {
                String[] ids = componentIds[i].split(":");
                if (ids.length == 2) {
                    try {
                        action.addCommand(new Action.Command(Integer.parseInt(ids[0]), Integer.parseInt(ids[1]), Integer.parseInt(operations[i]), getCommandParameter(parameters1, i), getCommandParameter(parameters2, i), getCommandParameter(parameters3, i)));
                    } catch (NumberFormatException e) {
                        return false;
                    }
                } else {
                    return false;
                }
            }
            actionsConcurrentMap.put(action.getId(), action);
            database.commit();
            return true;
        }
        return false;
    }

    protected boolean createDimmerAction(String description, String[] aliases, String[] idsAndChannel) {
        // TODO: Improve error handling, logging and user feedback
        if (description != null && description.trim().length() > 0 && aliases != null && aliases.length > 0 && idsAndChannel != null && idsAndChannel.length > 0) {
            // TODO: Improve action ID generation:
            int actionId = actionsConcurrentMap.size() + 1;
            while (actionsConcurrentMap.containsKey(actionId)) {
                actionId++;
            }
            description = description.trim();
            Action action = new Action(Action.ActionType.Dimmer, actionId, description);

            actionAliasesConcurrentMap.put(description.toLowerCase(), actionId);
            for (int i = 0; i < aliases.length; i++) {
                aliases[i] = aliases[i].trim();
                if (aliases[i].length() > 0) {
                    actionAliasesConcurrentMap.put(aliases[i].toLowerCase(), actionId);
                    action.addAlias(aliases[i]);
                }
            }
            for (int i = 0; i < idsAndChannel.length; i++) {
                String[] ids = idsAndChannel[i].split(":");
                if (ids.length == 3) {
                    try {
                        action.addCommand(new Action.Command(Integer.parseInt(ids[0]), Integer.parseInt(ids[1]), 0x0031, Integer.parseInt(ids[2]), null, null));
                    } catch (NumberFormatException e) {
                        return false;
                    }
                } else {
                    return false;
                }
            }
            actionsConcurrentMap.put(actionId, action);

            database.commit();
            return true;
        }
        return false;
    }

    protected Integer getCommandParameter(String[] parameters, int i) {
        Integer parameter = null;
        if (i < parameters.length && parameters[i] != null && !"".equals(parameters[i])) {
            try {
                parameter = Integer.parseInt(parameters[i].trim());
            } catch (NumberFormatException e) {
                logger.error("Got NumberFormatException while tyring to parse [" + parameters[i] + "] to an integer. Ignoring problem and not adding parameter to command.", e);
            }
        }
        return parameter;
    }

    protected ModelAndView getModelAndView(Request request, String page) {
        Map<String, Object> model = new HashMap<>();
        model.put("page", page);
        model.put("config", configConcurrentMap);
        if (hdlComponents == null || componentsConcurrentMap.size() != hdlComponents.size()) {
            hdlComponents = new ArrayList<>(componentsConcurrentMap.values());
            hdlComponents.sort((left, right) -> left.getSubnet() != right.getSubnet() ? left.getSubnet() - right.getSubnet() : left.getDeviceId() - right.getDeviceId());
        }
        model.put("components", hdlComponents);
        model.put("actions", actionsConcurrentMap.values());
        model.put("host", request.host());
        model.put("successMessage", StringEscapeUtils.escapeHtml(request.queryParams("success")));
        model.put("errorMessage", StringEscapeUtils.escapeHtml(request.queryParams("error")));
        model.put("HdlUtil", HdlUtil.class);

        return new ModelAndView(model, "templates/" + page + ".vm");
    }

    protected void requireSiteLocalAddress(Request request) {
        try {
            InetAddress remoteInetAddress = InetAddress.getByName(request.ip());
            if (!remoteInetAddress.isSiteLocalAddress()) {
                if (DEVEL_MODE) {
                    logger.warn("Running in development mode - allowing request from non-local IP address [" + request.ip() + "].");
                } else {
                    logger.warn("Access denied for resource [" + request.url() + "] for IP address [" + request.ip() + "]. User agent is [" + request.userAgent() + "].");
                    halt(401, "Access denied. Only allowing access from local addresses.");
                }
            }
        } catch (UnknownHostException e) {
            logger.error("Got UnknownHostException while trying to resolve address [" + request.ip() + "]. Halting request.");
            halt(401, "Access denied. Unable to verify that request is coming from a local address.");
        }
    }

    protected void requireValidAuthToken(String incomingAuthToken) {
        if (incomingAuthToken == null || incomingAuthToken.length() == 0 || !configConcurrentMap.get("authToken").equals(incomingAuthToken)) {
            halt(401, "Invalid authentication token.");
        }
    }

}
